### PR TITLE
[#653] styling timeline

### PR DESCRIPTION
**Pull request by abacz**
## What is in this PR?

Styled and functional timeline

### CODE REVIEW

File: s4e-web/src/app/views/map-view/state/product/product.service.ts [hunk 1]
```diff
@@ -37,21 +50,46 @@ export class ProductService {
...
+
+    if (product == null) {
+      return;
+    }
+
+    this.store.ui.update(product.id, state => ({...state, isLoading: true}));
+    this.http.get<MostRecentScene>(`${environment.apiPrefixV1}/products/${product.id}/scenes/most-recent`,
+      {params: {timeZone: timezone()}}
+    ).pipe(
+      finalize(() => this.store.ui.update(product.id, state => ({...state, isLoading: false})))
```
Comment (danielkryska) [inline s4e-web/src/app/views/map-view/state/product/product.service.ts:64]: Why haven't you used handleHttpRequest$ operator? Why do you not handle errors in this pipe?

File: s4e-web/src/app/views/map-view/state/product/product.service.ts [hunk 2]
```diff
@@ -37,21 +50,46 @@ export class ProductService {
...
       this.getSingle$(product)
-        .pipe(finalize(() => this.store.ui.update(productId, {isLoading: false})))
-        .subscribe((product) => {
-          applyTransaction(() => {
-            this.store.update(state => ({...state, ui: {...state.ui, loadedMonths: [], availableDays: []}}));
-            this.store.setActive(productId);
-          });
-          this.getAvailableDays();
+        .pipe(
+          finalize(() => this.store.ui.update(productId, {isLoading: false}))
```
Comment (danielkryska) [inline s4e-web/src/app/views/map-view/state/product/product.service.ts:85]: Why haven't you used handleHttpRequest$ operator? Why do you not handle errors in this pipe?

File: s4e-web/src/app/views/map-view/state/scene/scene.query.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import {SceneState, SceneStore} from './scene.store.service';
+import {Scene, SceneWithUI} from './scene.model';
+import {combineLatest, Observable} from 'rxjs';
+import {map} from 'rxjs/operators';
+import {ProductQuery} from '../product/product.query';
+import moment from 'moment';
+// import {combineLatest, Observable, of} from 'rxjs';
+// import {ICompleteRecentView} from '../recent-view/recent-view.model';
+// import {distinctUntilChanged, map, mergeMap} from 'rxjs/operators';
+// import {IUILayer} from '../common.model';
```
Comment (danielkryska) [inline s4e-web/src/app/views/map-view/state/scene/scene.query.ts:12]: Why are here those commented out imports?

File: s4e-web/src/app/views/map-view/state/scene/scene.query.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  //     map(views => views.map(view => ({
+  //       ...view,
+  //       activeProduct: this.productQuery.getEntity(view.productId),
+  //       activeProductType: this.productTypeQuery.getEntity(view.productTypeId)
+  //     })))
+  //   );
+  // }
+  //
+  // selectActiveViewProducts(): Observable<Scene[]> {
+  //   return this.selectActive().pipe(
```
Comment (danielkryska) [inline s4e-web/src/app/views/map-view/state/scene/scene.query.ts:69]: Why are here those commented out methods?

File: s4e-web/src/app/views/map-view/state/scene/scene.service.ts [hunk 5]
```diff
@@ -4,34 +4,49 @@ import {Injectable} from '@angular/core';
...
+        this.store.set(scenes);
+        if (scenes.length > 0) {
+          if (setActive === 'last') {
+            this.store.setActive(scenes[scenes.length - 1].id);
+          } else if (setActive === 'first') {
+            this.store.setActive(scenes[0].id);
+          }
+        } else {
+          this.store.setActive(null);
+        }
```
Comment (danielkryska) [inline s4e-web/src/app/views/map-view/state/scene/scene.service.ts:48]: Wouldn't it be simple to write:

```
const activeSceneId = null; 
if (scenes.length > 0) {
  activeSceneId = setActive === 'last' && scenes.length - 1 
    || setActive === 'first' && scenes[0].id
    || null;
} 
this.store.setActive(activeSceneId);

File: s4e-web/src/app/views/map-view/timeline/timeline.component.ts [hunk 6]
```diff
@@ -126,5 +145,17 @@ export class TimelineComponent implements OnInit, OnDestroy {
...
   }
 
-  ngOnDestroy(): void {}
+  ngOnDestroy(): void {
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    if (changes.resolution || changes.startTime) {
+      this.hourmarks = [];
+      for (let i = 0; i < this.resolution && this.hourmarks.length < TimelineComponent.HOURMARKS_COUNT; i += (this.resolution / TimelineComponent.HOURMARKS_COUNT)) {
```
Comment (danielkryska) [inline s4e-web/src/app/views/map-view/timeline/timeline.component.ts:154]: Wouldn't it be more efficient to store this equitation in variable `this.resolution / TimelineComponent.HOURMARKS_COUNT`?

**Review by danielkryska — APPROVED** [review]

**Review by jswk — APPROVED** [review]
I think the new timeline in general works and looks fine.
The information shown on no scenes is good, and the link is helpful to go to actual data.
I have a few remarks, which don't affect my approval.

At first, I had a problem locating the timeline. I thought it didn't show up or something, only then I realized it's white, and the contrast could be somehow improved.
image

Another thing visible above is the behaviour of zoom. I'd expect it to work analogously to map zoom, i.e. clicking + would constrain the range, and - would enlarge it. I needed a moment to adapt to these reversed functions.

**Comment by abacz** [comment]
> At first, I had a problem locating the timeline. I thought it didn't show up or something, only then I realized it's white, and the contrast could be somehow improved.

Yeap. On the openstreetmap is not visible immediately. I am waiting for functionality where newest scene will appear  when user clicks product, so timeline will be more visible, maybe. I need to check after implementation.
So it is good time to do this in nearest sprint. ;)

**Pull request merged**